### ISSUE TITLE
Explain and document the (gRPC's own) wire compatibility and the rules for gRPC libraries version bumps

### ISSUE THREAD

**Issue by MOZGIII** [grpc/grpc#17352]
Disclamer: I tried to google for it, coundn't find anything relevant; if it is already documented somewhere, there probably should be a link in the README.md.
 
### What version of gRPC and what language are you using?
gRPC 1.10 with Java, Go, Node.js, Web, to name a few.

Other predefined fields are irrelevant for this issue.

Consider this situation:
- There are a bunch of services in various languages/runtimes, let it be Go, Java and JavaScript/Node.js. All of them are using the same gRPC version (let it be 1.10.1).
- We want to update the gRPC to 1.16.x for all of the services. However, suppose it's pretty difficult to update all of the services at once. (This is easy to image if one's not using a monorepo.)
- It is easy, however, to update a single service to a newer gRPC version.

Now, the problem is: there's no clear way to understand upfront, if the setup with some services on gRPC 1.10.x and others on 1.16.x is expected to work. And the reason is there's no wire format compatibility promise built in the gRPC libraries versionning schema. Or, if there is - it's just out somewhere, maybe in just someone's head, maybe in the mailing list - and it is not actaully documented properly.

I this this problem has to be addressed by the project team. As a simplest solution, for now, I'd propose bumping major version for any backward-incompatible change in the wire format *or* generated code, in addition to other semver rules that projects follow. If that's stated explitily there should be no further undeterminism in this area.

To improve this futher, the compatibility can even be auto-tested across different minor/patch library versions of the same major version. Like Go gRPC 1.10.1 client conntecting to Go gRPC server 1.16.4. Or maybe even cross-library: Java gRPC 1.10.1 client connecting to Go gRPC server 1.16.4.

PS: The closest thing I could find is the issue #7530 (although is seems to be dead). Here I have a concrete and simple proposal, so maybe that issue can be closed in favor of this one.

**Comment by ericgribkoff** [comment]
You should have no issues with 1.10.1 and 1.16.x interoperability between languages. We use semantic versioning, so all 1.x releases should be compatible. 
gRPC-Go explicitly documents that they use semantic versioning here: https://github.com/grpc/grpc-go/blob/master/Documentation/versioning.md. The underlying wire format over HTTP2 (gRPC-web uses a separate protocol) is defined here: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md. 

(#7530 is, I think, slightly different - that's about changes impacting the underlying wire format, not versioning of the gRPC libraries themselves)

I agree that the versioning scheme should be called out here in the core repo as well. cc @hsaliak  @carl-mastrangelo

**Comment by carl-mastrangelo** [comment]
@MOZGIII FWIW, we do test compatibility across pretty much every version of gRPC and across languages and OSes.   It's not super discoverable, but it is public:

https://github.com/grpc/grpc/blob/master/tools/interop_matrix/client_matrix.py


Where would you expect the compatibility to be documented?    We can put a snippet wherever that is so that it's more discover-able in the future.

**Comment by MOZGIII** [comment]
Regarding this issue and #7530 - my point in this issue is all libraries (including you example with go library) only describe the *API* compatibility. Not a single word there about the wire format.

> The gRPC-Go versioning policy follows the Semantic Versioning 2.0.0 specification, with the following exceptions:
> - A MINOR version will not necessarily add new functionality.
> - MINOR releases will not break backward compatibility, except in the following circumstances:
>   - An API was marked as EXPERIMENTAL upon its introduction.
>   - An API was marked as DEPRECATED in the initial MAJOR release.
>   - An API is inherently flawed and cannot provide correct or secure behavior.
>   In these cases, APIs MAY be changed or removed without a MAJOR release. Otherwise, backward compatibility will be preserved by MINOR releases.
>   For an API marked as DEPRECATED, an alternative will be available (if appropriate) for at least three months prior to its removal.

I read this specifications as API compatibility guarantee, but not as a wire level compatibility guarantee. What if this implies that API will not change, but the underlying gRPC protocol may change, though in a way that if you bump all the libraries to that same updated version? *The documentation does not give me answer,* but I think it should.
I mean, maybe it's just me, and it's obvious for everyone else?

So, is it correct that current wire format versioning policy means any 1.x release is, on the wire-level, compatible with any other 1.x release? Seems like it's the case from https://github.com/grpc/grpc/blob/master/tools/interop_matrix/client_matrix.py (kudos for pointing me to that file, very helpful).

If it is, let's just tell people about this. In the docs, in the readme, just one more line and we're good.

**Comment by nicolasnoble** [comment]
From the wire protocol's perspective, it's fully defined and described here: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

This wire protocol got implemented even before grpc got public, and hasn't changed, across any of the implementations, even when we were in alpha. Even if we were to do any major semantic version of grpc, it may very well be that the version bump would be on API specification levels, and that the wire protocol still be compatible with the former major version.

In other words, the protocol documentation itself isn't versioned, because we have no plans on changing it, ever. It wouldn't be called grpc anymore otherwise.

Does this answer your question and concern?

**Comment by MOZGIII** [comment]
@nicolasnoble yes. Well, partially, because my concern is also about this not being documented properly.
If the story of origin and the view of the future that you just told here was in the preface at the protocol specification docs (at https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) - it would be a lot easier to understand what to expect from the protocol.
Maybe there should be not just one line, but a dedicated section on this topic in the readme/documentation website.
Also, a documentation on the concern boundaries (in terms of separation of concerns) could be improved. What is considered a part of a wire format, and what's not? For example, protobuf 3 had recently changed the decision on whether to preserve or not unknown fields in the messages. Can we still consider wire format compatible? What if the decision was changed in the opposite direction?

Also, is `gRPC over HTTP2` = `gRPC`?

**Comment by MOZGIII** [comment]
As a recap, the example I provided above is not correct.
The correct version would be:

> Note: every release version is compatible on the wire level with every other release version of gRPC, including different major versions and including across different language libraries. For more details, take a look at how we test the compatibility: https://github.com/grpc/grpc/blob/master/tools/interop_matrix/client_matrix.py

This captures the promise that gRPC is expected to be called something other than gRPC if the wire format changes. And that major version won't be used to encode that.

**Comment by MOZGIII** [comment]
> The above explanations I'm doing are part of the things we commonly explain to people, but I can't necessarily speak about what we're willing to set in stone in terms of specs.
Good point. I'd like to elaborate on that. The thing is, if this is a common question, and you explain that to people, it is better if it's specified somewhere. Now, what if it's not really the final design to be set in stone? Well, you're explaining it to people anyway, right? And it can potentially change in the future... Maybe it's better to have an ability to "version control" the design itself? I'm not talking about the formal process of the design iterations - clearly if there's no changes intended there's no need for the process - but at least dumping the current version of the de-facto position on the things we touched here is better than not doing it - I explained this point of mine before. Now, writing something is the docs does not necessarily mean it's set "in stone". In can be changed if needed, but the change will at least be transparent, and explicit, and people will notice.
It's important, of course, to specify just enough to bring the value, but to not add unwanted constraints on development of the project. In this case, though, I think the value is there, and there are no additional constraints.

**Comment by gnossen** [comment]
Bump. Do the involved parties feel that this conversation has led to anything actionable?
@MOZGIII @ericgribkoff @carl-mastrangelo @ejona86 @nicolasnoble

**Comment by MOZGIII** [comment]
I think we need to compose and add a small new section to the readme, stating that:
- gRPC wire format is compatible across all releases
- gRPC wire format will not change until explicit notice
- The statements above enforce by automatic tests which can be found at https://github.com/grpc/grpc/blob/master/tools/interop_matrix/client_matrix.py

That alone would render this issue solved for me.

**Comment by andrden** [comment]
In my experience (gRPC server in Java, android, iOS and gRPC-web clients) the cause of client-server incompatibility during the last year was another library used for securing gRPC: io.netty:netty-tcnative-boringssl-static, not gRPC itself
